profiler/src/Tools/ETW/dd-prof-etw/shared/IpcServer.cpp

```diff
@@ -0,0 +1,156 @@
+                     uint32_t inBufferSize,
+                     uint32_t outBufferSize,
+                     uint32_t maxInstances,
+                     uint32_t timeoutMS)
+{
+    _portName = portName;
+    _inBufferSize = inBufferSize;
+    _outBufferSize = outBufferSize;
+    _maxInstances = maxInstances;
+    _timeoutMS = timeoutMS;
+    _pHandler = pHandler;
+
+    _serverCount = 0;
+}
+
+std::unique_ptr<IpcServer> IpcServer::StartAsync(
+    bool showMessages,
+    const std::string& portName,
+    INamedPipeHandler* pHandler,
+    uint32_t inBufferSize,
+    uint32_t outBufferSize,
+    uint32_t maxInstances,
+    uint32_t timeoutMS
+    )
+{
+    if (pHandler == nullptr)
+    {
+        return nullptr;
+    }
+
+    auto server = std::make_unique<IpcServer>(
+        showMessages, portName, pHandler, inBufferSize, outBufferSize, maxInstances, timeoutMS
+        );
+
+    // let a threadpool thread process the command; allowing the server to process more incoming commands
+    if (!::TrySubmitThreadpoolCallback(StartCallback, (PVOID)server.get(), nullptr))
+    {
+        server->ShowLastError("Impossible to add the Start callback into the threadpool...");
+        return nullptr;
+    }
+
+    return server;
+ }
+
+void IpcServer::Stop()
+{
+    _stopRequested.store(true);
+}
+
+void CALLBACK IpcServer::StartCallback(PTP_CALLBACK_INSTANCE instance, PVOID context)
+{
+    IpcServer* pThis = reinterpret_cast<IpcServer*>(context);
+
+    // TODO: there is no timeout on ConnectNamedPipe
+    // so we would need to use the overlapped version to support _stopRequested :^(
+    while (!pThis->_stopRequested.load())
+    {
+        HANDLE hNamedPipe =
+            ::CreateNamedPipeA(
+                pThis->_portName.c_str(),
+                PIPE_ACCESS_DUPLEX,
+                PIPE_TYPE_MESSAGE | PIPE_READMODE_MESSAGE | PIPE_WAIT | PIPE_REJECT_REMOTE_CLIENTS,
+                pThis->_maxInstances,
+                pThis->_outBufferSize,
+                pThis->_inBufferSize,
+                0,
+                nullptr
+                );
+
+        pThis->_serverCount++;
+
+        if (hNamedPipe == INVALID_HANDLE_VALUE)
+        {
+            pThis->ShowLastError("Failed to create named pipe...");
+            if (pThis->_showMessages)
+            {
+                std::cout << "--> for server #" << pThis->_serverCount << "...\n";
+            }
+
+            pThis->_pHandler->OnStartError();
+            return;
+        }
+
+        if (pThis->_showMessages)
+        {
+            std::cout << "Listening to server #" << pThis->_serverCount << "...\n";
+        }
+
+        if (!::ConnectNamedPipe(hNamedPipe, nullptr) && ::GetLastError() != ERROR_PIPE_CONNECTED)
+        {
+            pThis->ShowLastError("ConnectNamedPipe failed...");
+            ::CloseHandle(hNamedPipe);
+
+            pThis->_pHandler->OnConnectError();
+            return;
+        }
+
+        // TODO: how to use smart pointers in that case?
```

Comment:
In this case, it would not be interesting to use a smart pointer. At best, the code will be difficult to read.